### PR TITLE
Make the TransportRolloverAction execute in one cluster state update

### PR DESCRIPTION
**Pull request by andreidan**
This commit makes the TransportRolloverAction more resilient, by having it 
execute only one cluster state update that creates the new (rollover index), rolls 
over the alias from the source to the target index and set the RolloverInfo on the
source index. Before, these 3 steps were represented as 3 chained cluster state
updates, which would've seen the user manually intervene if, say, the alias
rollover cluster state update (second in the chain) failed but the creation of
the rollover index (first in the chain) update succeeded

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java [hunk 1]
```diff
@@ -141,56 +140,41 @@ public void onResponse(IndicesStatsResponse statsResponse) {
...
-                            final IndicesAliasesClusterStateUpdateRequest aliasesUpdateRequest;
-                            if (explicitWriteIndex) {
-                                aliasesUpdateRequest = prepareRolloverAliasesWriteIndexUpdateRequest(sourceIndexName,
-                                    rolloverIndexName, rolloverRequest);
-                            } else {
-                                aliasesUpdateRequest = prepareRolloverAliasesUpdateRequest(sourceIndexName,
-                                    rolloverIndexName, rolloverRequest);
+                        CreateIndexClusterStateUpdateRequest createIndexRequest = prepareCreateIndexRequest(unresolvedName,
+                            rolloverIndexName, rolloverRequest);
+                        clusterService.submitStateUpdateTask("rollover_index", new ClusterStateUpdateTask() {
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java]: Can you add the index name into the task string?

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java [hunk 2]
```diff
@@ -141,56 +140,41 @@ public void onResponse(IndicesStatsResponse statsResponse) {
...
-                                        });
-                                    } else {
-                                        listener.onResponse(new RolloverResponse(sourceIndexName, rolloverIndexName, conditionResults,
-                                                                                    false, true, false, false));
-                                    }
-                                }, listener::onFailure));
-                        }, listener::onFailure));
+                            @Override
+                            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                                activeShardsObserver.waitForActiveShards(new String[]{rolloverIndexName},
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java]: I don't expect it to happen (it would be a bad situation if it did), but it might be a good idea to wrap this in:

```java
if (newState.equals(oldState) == false) {
    ...
}
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java [hunk 3]
```diff
@@ -30,6 +30,11 @@ public RolloverStep(StepKey key, StepKey nextStepKey, Client client) {
         super(key, nextStepKey, client);
     }
 
+    @Override
+    public boolean isRetryable() {
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java]: I think we should do this in a separate PR. We still need to solve the complexity around retrying `AsyncAction` steps, which will require slightly different execution than the other step types, due to their exactly-once re-invocation.

**Review by dakrone — CHANGES_REQUESTED** [review]
Thanks @andreidan, I left two minor comments and one about moving part of this to a separate PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java (hunk 3 above)
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java]: I think we satisfy the exactly-once re-invocation already. A failed `AsyncActionStep` will be moved into the `ErrorStep` on failure and on the next ILM periodic loop we'll move the index policy back into the failed step (the `AsyncActionStep`) which will be executed on the next async steps execution cycle (on a master change event or the subsequent periodic ILM loop). Am I missing something here? I'm happy to separate the PRs either way

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java (hunk 3 above)
Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java]: `AsyncActionStep`s don't get executed by the periodic loop invocation, meaning they will never be executed except on a master change event (which should be exceedingly rare). I was saying that we need to figure out the best way to make sure we can execute `AsyncAction` steps when they do an automatic retry.

To figure out this logic I was thinking a different PR would be better since it would likely mean extra testing for however we decide we want invoke `AsyncAction` steps (it may be as simple as adding the invocation in the `onProcessed` method of the `ilm-retry-failed-step` cluster state update, but we'll have to test).

**Review by dakrone — APPROVED** [review]
LGTM, thanks Andrei

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java (hunk 3 above)
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RolloverStep.java]: Ah, you're right Lee, I mixed the `AsyncWaitStep` with the `AsyncActionStep` (async and step was all my brain read)

**Comment by shwetathareja** [comment]
Hi @andreidan, 
Do you have plans to backport this to 6.8? We are facing similar problem in our cluster as pointed out here [issue](https://github.com/elastic/elasticsearch/issues/44135) and failing to rollover is causing single index to grow huge. 
If it is ok, I can raise the PR for backporting to 6.8.

**Comment by andreidan** [comment]
Hi @shwetathareja, thank you for your interest and I'm sorry to hear you're running into issues with the rollover action.
Unfortunately, we will not backport this to 6.8 as that release line will only receive bug fixes going forward. We are working extensively on making ILM more reliable (with this particular enhancement backported already to 7.6 and more to come with regards to making ILM and the rollover action more resilient). Would an upgrade to 7.6 be possible once it's out?
